ga4gh/gff3Parser.py

```diff
@@ -0,0 +1,321 @@
+"""
+Object representation of GFF3 data and parser for GFF3 files.
+
+See: http://www.sequenceontology.org/gff3.shtml
+"""
+
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import urllib
+import copy
+import re
+import gzip
+import bz2
+import collections
+
+GFF3_HEADER = "##gff-version 3"
+
+
+class GFF3Exception(Exception):
+    """
+    Exception associated with GFF3 data.
+    """
+    def __init__(self, message, fileName=None, lineNumber=None):
+        if fileName is not None:
+            message = "{}:{}: {}".format(fileName, lineNumber, message)
+        super(GFF3Exception, self).__init__(message)
+
+# characters forcing encode for columns
+_encodeColReStr = "\t|\n|\r|%|[\x00-\x1F]|\x7f"
+_encodeColRe = re.compile(_encodeColReStr)
+
+# characters forcing encode for attribute names or values
+_encodeAttrReStr = _encodeColReStr + "|;|=|&|,"
+_encodeAttrRe = re.compile(_encodeAttrReStr)
+
+
+def _encodeCol(v):
+    """
+    Encode a column if is has special characters.
+    """
+    if _encodeColRe.search(v):
+        return urllib.quote(v)
+    else:
+        return v
+
+
+def _encodeAttr(v):
+    """
+    Encode a attribute name or value if is has special characters.
+    """
+    if _encodeAttrRe.search(v):
+        return urllib.quote(v)
+    else:
+        return v
+
+
+class Feature(object):
+    """
+    One feature notation from a GFF3, missing attribute, as code by `.' in
+    the file are stored as None.
+    """
+
+    __slots__ = ("seqname", "source", "type", "start", "end", "score",
```

Comment:
Either delete it or add a comment; I don't see the down side of
having it, but I don't care.  I put in in there because I was
process very larger files.  It was not worth my time to do
performance analysis to see if it mattered, as there were no
downsides to using a standard python feature.


Maciek Smuga-Otto <notifications@github.com> writes:
> Put me on the record for agreeing wholeheartedly with @jeromekelleher about
> premature optimization. We can revisit this once we've hit an actual issue with
> use.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub*
>